src/backends/plonky2/primitives/ec/curve.rs

```diff
@@ -0,0 +1,759 @@
+            u: FieldTarget::new(src.read_target_array()?),
+        };
+        let sqrt = PointTarget {
+            x: FieldTarget::new(src.read_target_array()?),
+            u: FieldTarget::new(src.read_target_array()?),
+        };
+        Ok(Self { orig, sqrt })
+    }
+}
+
+pub trait CircuitBuilderElliptic {
+    fn add_virtual_point_target(&mut self) -> PointTarget;
+    fn identity_point(&mut self) -> PointTarget;
+    fn constant_point(&mut self, p: Point) -> PointTarget;
+
+    fn add_point(&mut self, p1: &PointTarget, p2: &PointTarget) -> PointTarget;
+    fn double_point(&mut self, p: &PointTarget) -> PointTarget;
+    fn linear_combination_points(
+        &mut self,
+        p1_scalar: &[BoolTarget; 320],
+        p2_scalar: &[BoolTarget; 320],
+        p1: &PointTarget,
+        p2: &PointTarget,
+    ) -> PointTarget;
+    fn if_point(
+        &mut self,
+        b: BoolTarget,
+        p_true: &PointTarget,
+        p_false: &PointTarget,
+    ) -> PointTarget;
+
+    /// Check that two points are equal.  This assumes that the points are
+    /// already known to be in the subgroup.
+    fn connect_point(&mut self, p1: &PointTarget, p2: &PointTarget);
+    fn check_point_on_curve(&mut self, p: &PointTarget);
+    fn check_point_in_subgroup(&mut self, p: &PointTarget);
+}
+
+impl CircuitBuilderElliptic for CircuitBuilder<GoldilocksField, 2> {
+    fn add_virtual_point_target(&mut self) -> PointTarget {
+        let p = PointTarget {
+            x: self.add_virtual_nnf_target(),
+            u: self.add_virtual_nnf_target(),
+        };
+        self.check_point_in_subgroup(&p);
+        p
+    }
+
+    fn identity_point(&mut self) -> PointTarget {
+        self.constant_point(Point::ZERO)
+    }
+
+    fn constant_point(&mut self, p: Point) -> PointTarget {
+        assert!(p.is_in_subgroup());
+        PointTarget {
+            x: self.nnf_constant(&p.x),
+            u: self.nnf_constant(&p.u),
+        }
+    }
+
+    fn add_point(&mut self, p1: &PointTarget, p2: &PointTarget) -> PointTarget {
+        let mut inputs = Vec::with_capacity(20);
+        inputs.extend_from_slice(&p1.x.components);
+        inputs.extend_from_slice(&p1.u.components);
+        inputs.extend_from_slice(&p2.x.components);
+        inputs.extend_from_slice(&p2.u.components);
+        let mut outputs = ECAddHomog::apply(self, &inputs);
+        // plonky2 expects all gate constraints to be satisfied by the zero vector.
+        // So our elliptic curve addition gate computes [x,z-b,u,t-b], and we have to add the b here.
+        let b1 = self.constant(Point::B1);
+        outputs[6] = self.add(outputs[6], b1);
+        outputs[16] = self.add(outputs[16], b1);
+        let x = FieldTarget::new(outputs[0..5].try_into().unwrap());
+        let z = FieldTarget::new(outputs[5..10].try_into().unwrap());
+        let u = FieldTarget::new(outputs[10..15].try_into().unwrap());
+        let t = FieldTarget::new(outputs[15..20].try_into().unwrap());
```

Comment:
```suggestion
        let x = FieldTarget::new(outputs[0..5].try_into().unwrap());
        let z = FieldTarget::new(outputs[5..10].try_into().unwrap());
        let u = FieldTarget::new(outputs[10..15].try_into().unwrap());
        let t = FieldTarget::new(outputs[15..20].try_into().unwrap());
        let b1 = self.constant(Point::B1);
        let z = self.nnf_add_scalar_times_generator_power(b1, 0, z);
        let t = self.nnf_add_scalar_times_generator_power(b1, 0, t);
```
This is a suggestion that may make the code a bit more clear, as we don't need indexing into `outputs`.  Feel free to ignore.